pkg/controller/elasticsearch/driver/upgrade_deletion_strategy.go

```diff
@@ -0,0 +1,263 @@
+		}
+		if label.IsMasterNode(pod1) && !label.IsMasterNode(pod2) {
+			// pod2 has higher priority since it is a data node
+			return false
+		}
+		return true
+	})
+}
+
+var predicates = [...]Predicate{
+	{
+		// If MaxUnavailable is reached, allow for an unhealthy Pod to be deleted.
+		// This is to prevent a situation where MaxUnavailable is reached and we
+		// can't make some progress even if the user has updated the spec.
+		name: "do_not_restart_healthy_node_if_MaxUnavailable_reached",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			if maxUnavailableReached && k8s.IsPodReady(candidate) {
+				return false, nil
+			}
+			return true, nil
+		},
+	},
+	{
+		name: "skip_unknown_or_long_terminating_pods",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			if candidate.DeletionTimestamp != nil && candidate.Status.Reason == NodeUnreachablePodReason {
+				// kubelet is unresponsive, Unknown Pod, do not try to delete it
+				return false, nil
+			}
+			if candidate.DeletionTimestamp != nil && time.Now().After(candidate.DeletionTimestamp.Add(TerminatingPodTimeout)) {
+				return false, nil
+			}
+			return true, nil
+		},
+	},
+	{
+		// In Yellow or Red status only allow unhealthy Pods to be restarted.
+		// This is intended to unlock some situations where the cluster is not green and
+		// a Pod has to be restarted a second time.
+		name: "do_not_restart_healthy_node_if_not_green",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			green, err := context.esState.GreenHealth()
+			if err != nil {
+				return false, err
+			}
+			if green {
+				return true, nil
+			}
+			if !k8s.IsPodReady(candidate) {
+				return true, nil
+			}
+			return false, nil
+		},
+	},
+	{
+		// One master at a time
+		name: "one_master_at_a_time",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			// If candidate is not a master then we don't care
+			if !label.IsMasterNode(candidate) {
+				return true, nil
+			}
+			// If candidate is not healthy we want give it a chance to restart
+			_, healthy := context.healthyPods[candidate.Name]
+			if !healthy {
+				return true, nil
+			}
+
+			for _, pod := range deletedPods {
+				if label.IsMasterNode(pod) {
+					return false, nil
+				}
+			}
+			// Get the expected masters
+			expectedMasters := len(context.masterNodesNames)
+			// Get the healthy masters
+			healthyMasters := 0
+			for _, pod := range context.healthyPods {
+				if label.IsMasterNode(pod) {
+					healthyMasters++
+				}
+			}
+			return healthyMasters == expectedMasters, nil
```

Comment:
Maybe add a comment here explaining the subtle invariant that we can assume no upscale/downscale operation has added or removed a master node if the actual number of masters and the expected ones align. 

Also I think this relies on us not copying the resources list anywhere on the way down to here, so that any adjustments made in any of the up/downscale functions are actually visible.